### PR TITLE
feat(admin): user roles are split by source

### PR DESCRIPTION
**Pull request by Kuliak**
* user-detail page splits his roles by source (directly assigned or obtained as a member of authorized group)
* tabs added to change what source we want to see
* authorized-group-based roles cannot be managed, only viewed

### CODE REVIEW

**Review by Johaney-s — COMMENTED** [review]
It seems myProfile -> roles view is not working properly yet:
image

and just a thought, but it might be too much work for such little thing, showing which authoritative group is giving us the role could be useful.

File: apps/admin-gui/src/assets/i18n/en.json [hunk 1]
```diff
@@ -2511,7 +2511,9 @@
       "SHOW_RECENTLY_VIEWED": "Show Recently viewed",
       "SHOW_ROLES": "Show roles",
       "MAIL_CHANGE_SUCCESS": "Preferred mail has been changed",
-      "ANONYMIZED": "User anonymized"
+      "ANONYMIZED": "User anonymized",
+      "DIRECT_ROLES": "Direct",
+      "AUTHORIZED_GROUP_BASED_ROLES": "Authorized-group-based"
```
Comment (xflord) [inline apps/admin-gui/src/assets/i18n/en.json:2516]: Would prefer it without the dashes

**Review by xflord — APPROVED** [review]
Remove the now unused imports, otherwise looks good, good job

File: libs/perun/services/src/lib/role.service.ts [hunk 2]
```diff
@@ -21,4 +22,28 @@ export class RoleService {
     });
     return preparedRoles;
   }
+
+  prepareComplementaryObjects(
```
Comment (xflord) [inline libs/perun/services/src/lib/role.service.ts:26]: please add javadocs

File: apps/admin-gui/src/app/shared/components/roles-list/roles-page.component.html [hunk 3]
```diff
@@ -43,9 +43,11 @@ <h1 class="page-subtitle d-flex">{{'ROLES.TITLE'| translate}}</h1>
               {{('ROLES.' + role.roleName + '_VOS_' + entityType) | translate}}
             </div>
             <perun-web-apps-vos-list
-              [displayedColumns]="role.roleName === 'MEMBERSHIP' ? ['id', 'name', 'shortName'] : ['checkbox', 'id', 'name', 'shortName']"
+              [displayedColumns]="role.roleName === 'MEMBERSHIP' || !editable? ['id', 'name', 'authzGroup', 'shortName'] : ['checkbox', 'id', 'name', 'authzGroup', 'shortName']"
               [vos]="vos | async"
-              [selection]="selection">
+              [selection]="selection"
+              [authzVoNames]="voNames"
+              [voWithAuthzGroupPairs]="_complementaryObjectsWithAuthzGroups.get(role.roleName).get('vo')">
```
Comment (xflord) [inline apps/admin-gui/src/app/shared/components/roles-list/roles-page.component.html:50]: This causes exceptions in the user based tab.

**Review by xflord — CHANGES_REQUESTED** [review]

File: apps/admin-gui/src/app/shared/components/roles-list/roles-page.component.html [hunk 4]
```diff
@@ -27,6 +26,7 @@ <h1 class="page-subtitle d-flex">{{'ROLES.TITLE'| translate}}</h1>
         matTooltipPosition="left"
         [matTooltipDisabled]="(selection.selected.length === 0 && selectedFacilities.selected.length === 0) || !disableRemove">
         <button
+          *ngIf="editable"
```
Comment (bodnara) [inline apps/admin-gui/src/app/shared/components/roles-list/roles-page.component.html:29]: Nitpick but the proper target should be the parent of the button (span tag). Right now the span is present so the tooltip could be triggered.

File: apps/admin-gui/src/app/users/pages/user-detail-page/user-settings/user-roles/user-roles.component.ts [hunk 5]
```diff
@@ -43,22 +43,38 @@ export class UserRolesComponent implements OnInit {
   }
 
   getData(): void {
+    let loadingTypes = 3;
     this.outerLoading = true;
+
```
Comment (bodnara) [inline apps/admin-gui/src/app/users/pages/user-detail-page/user-settings/user-roles/user-roles.component.ts:48]: While this works a better solution (without rxjs magic) would be to nest these calls. Also stop the loading if some of the requests fail.

File: libs/perun/components/src/lib/facilities-list/facilities-list.component.html [hunk 6]
```diff
@@ -51,6 +51,16 @@
...
       </ng-container>
+      <ng-container matColumnDef="authzGroup">
+        <th mat-header-cell *matHeaderCellDef mat-sort-header>
+          {{'SHARED.COMPONENTS.FACILITIES_LIST.AUTHZ_GROUP' | translate}}
+        </th>
+        <td mat-cell *matCellDef="let facility">
+          {{this._facilityWithAuthzGroupPairs.has(facility.facility.id)
+              ? _authzVoNames.get(this._facilityWithAuthzGroupPairs.get(facility.facility.id)[0].voId) + ":" + this._facilityWithAuthzGroupPairs.get(facility.facility.id)[0].name
+              : "NO" }}
+        </td>
```
Comment (bodnara) [inline libs/perun/components/src/lib/facilities-list/facilities-list.component.html:62]: I prefer the empty string for the else clause.

File: libs/perun/components/src/lib/facilities-list/facilities-list.component.ts [hunk 7]
```diff
@@ -41,11 +41,21 @@ export class FacilitiesListComponent implements OnChanges {
...
 
   constructor(private authResolver: GuiAuthResolver) {}
 
+  @Input() set facilityWithAuthzGroupPairs(pairs: Map<number, Group[]>) {
+    this._facilityWithAuthzGroupPairs = pairs;
+  }
+
+  @Input() set authzVoNames(authzVoNames: Map<number, string>) {
+    this._authzVoNames = authzVoNames;
+  }
```
Comment (bodnara) [inline libs/perun/components/src/lib/facilities-list/facilities-list.component.ts:57]: It is not technically wrong to use input setter like this, it is more useful when there is a peace of code that needs to run every time the input changes. However to stay consistent with the rest of the codebase could you please make these a simple `@Input`. This goes for other list components.

**Review by bodnara — CHANGES_REQUESTED** [review]

File: apps/admin-gui/src/app/users/pages/user-detail-page/user-settings/user-roles/user-roles.component.ts [hunk 8]
```diff
@@ -44,21 +44,51 @@ export class UserRolesComponent implements OnInit {
...
+                    // prepare map
+                    this.rolesComplementaryObjectsWithAuthzGroups =
+                      this.roleService.prepareComplementaryObjects(
+                        Object.keys(obtainedComplementaryObjects),
+                        obtainedComplementaryObjects
+                      );
+                    this.outerLoading = false;
+                  },
+                  () => (this.outerLoading = false)
+                );
```
Comment (xflord) [inline apps/admin-gui/src/app/users/pages/user-detail-page/user-settings/user-roles/user-roles.component.ts:86]:
```suggestion
                .subscribe({
                  next: (obtainedComplementaryObjects) => {
                    // prepare map
                    this.rolesComplementaryObjectsWithAuthzGroups =
                      this.roleService.prepareComplementaryObjects(
                        Object.keys(obtainedComplementaryObjects),
                        obtainedComplementaryObjects
                      );
                    this.outerLoading = false;
                  },
                  error: () => (this.outerLoading = false)
                });
```
use the non-deprecated syntax with all the subscribe methods please (like this I believe)

**Review by xflord — COMMENTED** [review]
fix the scope to feat(admin) since the roles page is only in the admin gui

**Review by xflord — APPROVED** [review]

**Review by bodnara — COMMENTED** [review]
There is some uninitialized value in the group-list on user based tab.
image

**Review by bodnara — APPROVED** [review]

**Pull request merged**